File: src/convention/schema.ts
import type {
  ServiceIR,
  EntityDecl,
  FieldDecl,
  TypeExpr,
  Expr,
  TypeAliasDecl,
  EnumDecl,
  StateFieldDecl,
} from "#ir/types.js";
import type {
  DatabaseSchema,
  TableSpec,
  ColumnSpec,
  ForeignKeySpec,
  IndexSpec,
} from "#convention/types.js";
import { toTableName, toColumnName, toSnakeCase } from "#convention/naming.js";
import { getConvention } from "#convention/path.js";

export function deriveSchema(ir: ServiceIR): DatabaseSchema {
  const entityNames = new Set(ir.entities.map((e) => e.name));
  const enumMap = new Map(ir.enums.map((e) => [e.name, e]));
  const aliasMap = new Map(ir.typeAliases.map((a) => [a.name, a]));

  const tables: TableSpec[] = [];

  for (const entity of ir.entities) {
    tables.push(deriveTable(entity, ir, entityNames, enumMap, aliasMap));
  }

  if (ir.state) {
    for (const field of ir.state.fields) {
      if (field.typeExpr.kind !== "RelationType") continue;
      if (field.typeExpr.multiplicity === "some" || field.typeExpr.multiplicity === "set") {
        const junction = deriveJunctionTable(field, entityNames, enumMap, aliasMap);
        if (junction) tables.push(junction);
      }
    }
  }

  return { tables };
}

function deriveTable(
  entity: EntityDecl,
  ir: ServiceIR,
  entityNames: Set<string>,
  enumMap: Map<string, EnumDecl>,
  aliasMap: Map<string, TypeAliasDecl>,
): TableSpec {
  const tableNameOverride = getConvention(ir.conventions, entity.name, "db_table");
  const tableName = tableNameOverride || toTableName(entity.name);

  const entityFieldNames = new Set(entity.fields.map((f) => f.name));

  const columns: ColumnSpec[] = [];
  if (!entityFieldNames.has("id")) {
    columns.push({ name: "id", sqlType: "BIGSERIAL", nullable: false, defaultValue: null });
  }

  const foreignKeys: ForeignKeySpec[] = [];
  const checks: string[] = [];
  const indexes: IndexSpec[] = [];

  for (const field of entity.fields) {
    const colName = toColumnName(field.name);
    const mapped = mapFieldToColumn(field, entityNames, enumMap, aliasMap);

    columns.push(mapped.column);

    if (mapped.foreignKey) {
      foreignKeys.push(mapped.foreignKey);
      indexes.push({
        name: `idx_${tableName}_${colName}`,
        columns: [colName],
        unique: false,
      });
    }

    if (mapped.check) checks.push(mapped.check);

    if (field.constraint) {
      const constraintChecks = extractChecks(colName, field.constraint);
      checks.push(...constraintChecks);
    }
  }

  for (const inv of entity.invariants) {
    const invChecks = extractInvariantChecks(inv, entity.fields);
    checks.push(...invChecks);
  }

  if (ir.state) {
    for (const sf of ir.state.fields) {
      if (sf.typeExpr.kind !== "RelationType") continue;
      const toName = resolveTypeName(sf.typeExpr.toType);
      if (toName !== entity.name) continue;
      if (sf.typeExpr.multiplicity === "some" || sf.typeExpr.multiplicity === "set") continue;

      const fromName = resolveTypeName(sf.typeExpr.fromType);
      if (fromName && entityNames.has(fromName)) {
        const fkCol = toColumnName(fromName.replace(/([A-Z])/g, "_$1").toLowerCase().replace(/^_/, "")) + "_id";
        const fkRefTable = toTableName(fromName);
        const nullable = sf.typeExpr.multiplicity === "lone";

        if (!columns.some((c) => c.name === fkCol)) {
          columns.push({
            name: fkCol,
            sqlType: "BIGINT",
            nullable,
            defaultValue: null,
          });
          foreignKeys.push({
            column: fkCol,
            refTable: fkRefTable,
            refColumn: "id",
            onDelete: "CASCADE",
          });
          indexes.push({
            name: `idx_${tableName}_${fkCol}`,
            columns: [fkCol],
            unique: false,
          });
        }
      }
    }
  }

  const tsOverride = getConvention(ir.conventions, entity.name, "db_timestamps");
  const addTimestamps = tsOverride !== "false";

  if (addTimestamps && !columns.some((c) => c.name === "created_at")) {
    columns.push({ name: "created_at", sqlType: "TIMESTAMPTZ", nullable: false, defaultValue: "NOW()" });
  }
  if (addTimestamps && !columns.some((c) => c.name === "updated_at")) {
    columns.push({ name: "updated_at", sqlType: "TIMESTAMPTZ", nullable: false, defaultValue: "NOW()" });
  }

  return {
    name: tableName,
    entityName: entity.name,
    columns,
    primaryKey: "id",
    foreignKeys,
    checks,
    indexes,
  };
}

function deriveJunctionTable(
  stateField: StateFieldDecl,
  entityNames: Set<string>,
  enumMap: Map<string, EnumDecl>,
  aliasMap: Map<string, TypeAliasDecl>,
): TableSpec | null {
  if (stateField.typeExpr.kind !== "RelationType") return null;

  const fromName = resolveTypeName(stateField.typeExpr.fromType);
  const toName = resolveTypeName(stateField.typeExpr.toType);
  if (!fromName || !toName) return null;
  if (!entityNames.has(fromName) || !entityNames.has(toName)) return null;

  const fromTable = toTableName(fromName);
  const toTable = toTableName(toName);
  const tableName = `${fromTable}_${toTable}`;

  const fromCol = toColumnName(fromName.replace(/([A-Z])/g, "_$1").toLowerCase().replace(/^_/, "")) + "_id";
  const toCol = toColumnName(toName.replace(/([A-Z])/g, "_$1").toLowerCase().replace(/^_/, "")) + "_id";

  return {
    name: tableName,
    entityName: `${fromName}_${toName}`,
    columns: [
      { name: "id", sqlType: "BIGSERIAL", nullable: false, defaultValue: null },
      { name: fromCol, sqlType: "BIGINT", nullable: false, defaultValue: null },
      { name: toCol, sqlType: "BIGINT", nullable: false, defaultValue: null },
      { name: "created_at", sqlType: "TIMESTAMPTZ", nullable: false, defaultValue: "NOW()" },
    ],
    primaryKey: "id",
    foreignKeys: [
      { column: fromCol, refTable: fromTable, refColumn: "id", onDelete: "CASCADE" },
      { column: toCol, refTable: toTable, refColumn: "id", onDelete: "CASCADE" },
    ],
    checks: [],
    indexes: [
      {
        name: `idx_${tableName}_${fromCol}_${toCol}`,
        columns: [fromCol, toCol],
        unique: true,
      },
      { name: `idx_${tableName}_${fromCol}`, columns: [fromCol], unique: false },
      { name: `idx_${tableName}_${toCol}`, columns: [toCol], unique: false },
    ],
  };
}

interface MappedField {
  column: ColumnSpec;
  foreignKey: ForeignKeySpec | null;
  check: string | null;
}

function mapFieldToColumn(
  field: FieldDecl,
  entityNames: Set<string>,
  enumMap: Map<string, EnumDecl>,
  aliasMap: Map<string, TypeAliasDecl>,
): MappedField {
  const colName = toColumnName(field.name);
  const mapped = mapTypeToColumn(colName, field.typeExpr, entityNames, enumMap, aliasMap);

  if (mapped.foreignKey === null && colName.endsWith("_id")) {
    const prefix = colName.slice(0, -"_id".length);
    const targetEntity = [...entityNames].find((n) => toSnakeCase(n) === prefix);
    if (targetEntity !== undefined) {
      return {
        column: { ...mapped.column, sqlType: "BIGINT" },
        foreignKey: { column: colName, refTable: toTableName(targetEntity), refColumn: "id", onDelete: "CASCADE" },
        check: mapped.check,
      };
    }
  }

  return mapped;
}

function mapTypeToColumn(
  colName: string,
  typeExpr: TypeExpr,
  entityNames: Set<string>,
  enumMap: Map<string, EnumDecl>,
  aliasMap: Map<string, TypeAliasDecl>,
): MappedField {
  switch (typeExpr.kind) {
    case "NamedType": {
      const name = typeExpr.name;

      const sqlType = PRIMITIVE_TYPE_MAP.get(name);
      if (sqlType) {
        return { column: { name: colName, sqlType, nullable: false, defaultValue: null }, foreignKey: null, check: null };
      }

      const enumDecl = enumMap.get(name);
      if (enumDecl) {
        const values = enumDecl.values.map((v) => `'${escapeSqlString(v)}'`).join(", ");
        return {
          column: { name: colName, sqlType: "TEXT", nullable: false, defaultValue: null },
          foreignKey: null,
          check: `${colName} IN (${values})`,
        };
      }

      if (entityNames.has(name)) {
        const refTable = toTableName(name);
        return {
          column: { name: colName + "_id", sqlType: "BIGINT", nullable: false, defaultValue: null },
          foreignKey: { column: colName + "_id", refTable, refColumn: "id", onDelete: "CASCADE" },
          check: null,
        };
      }

      const alias = aliasMap.get(name);
      if (alias) {
        return mapTypeToColumn(colName, alias.typeExpr, entityNames, enumMap, aliasMap);
      }

      return { column: { name: colName, sqlType: "TEXT", nullable: false, defaultValue: null }, foreignKey: null, check: null };
    }

    case "OptionType": {
      const inner = mapTypeToColumn(colName, typeExpr.innerType, entityNames, enumMap, aliasMap);
      return {
        column: { ...inner.column, nullable: true },
        foreignKey: inner.foreignKey ? { ...inner.foreignKey } : null,
        check: inner.check,
      };
    }

    case "SetType":
      return { column: { name: colName, sqlType: "JSONB", nullable: false, defaultValue: "'[]'::jsonb" }, foreignKey: null, check: null };

    case "SeqType":
      return { column: { name: colName, sqlType: "JSONB", nullable: false, defaultValue: "'[]'::jsonb" }, foreignKey: null, check: null };

    case "MapType":
      return { column: { name: colName, sqlType: "JSONB", nullable: false, defaultValue: "'{}'::jsonb" }, foreignKey: null, check: null };

    case "RelationType":
      return { column: { name: colName, sqlType: "BIGINT", nullable: false, defaultValue: null }, foreignKey: null, check: null };
  }
}

const PRIMITIVE_TYPE_MAP: ReadonlyMap<string, string> = new Map([
  ["String", "TEXT"],
  ["Int", "INTEGER"],
  ["Float", "DOUBLE PRECISION"],
  ["Bool", "BOOLEAN"],
  ["Boolean", "BOOLEAN"],
  ["DateTime", "TIMESTAMPTZ"],
  ["Date", "DATE"],
  ["UUID", "UUID"],
  ["Decimal", "NUMERIC(19,4)"],
  ["Bytes", "BYTEA"],
  ["Money", "INTEGER"],
]);

function escapeSqlString(value: string): string {
  return value.replace(/'/g, "''");
}

function resolveTypeName(typeExpr: TypeExpr): string | null {
  if (typeExpr.kind === "NamedType") return typeExpr.name;
  return null;
}

function extractChecks(colName: string, constraint: Expr): string[] {
  const checks: string[] = [];
  visitConstraint(constraint, colName, checks);
  return checks;
}

function visitConstraint(expr: Expr, colName: string, checks: string[]): void {
  if (expr.kind === "BinaryOp" && expr.op === "and") {
    visitConstraint(expr.left, colName, checks);
    visitConstraint(expr.right, colName, checks);
    return;
  }

  if (expr.kind === "BinaryOp") {
    const check = tryMapComparison(expr, colName);
    if (check) checks.push(check);
    return;
  }

  if (expr.kind === "Matches" && expr.expr.kind === "Identifier") {
    checks.push(`${colName} ~ '${escapeSqlString(expr.pattern)}'`);
  }
}

function tryMapComparison(expr: Expr & { kind: "BinaryOp" }, colName: string): string | null {
  const sqlOps: Record<string, string> = {
    ">": ">",
    "<": "<",
    ">=": ">=",
    "<=": "<=",
    "=": "=",
    "!=": "!=",
  };

  const sqlOp = sqlOps[expr.op];
  if (!sqlOp) return null;

  if (isLenCall(expr.left) && isLiteral(expr.right)) {
    return `length(${colName}) ${sqlOp} ${literalValue(expr.right)}`;
  }

  if (isValueRef(expr.left) && isLiteral(expr.right)) {
    return `${colName} ${sqlOp} ${literalValue(expr.right)}`;
  }

  return null;
}

function isLenCall(expr: Expr): boolean {
  return (
    expr.kind === "Call" &&
    expr.callee.kind === "Identifier" &&
    expr.callee.name === "len"
  );
}

function isValueRef(expr: Expr): boolean {
  return expr.kind === "Identifier" && expr.name === "value";
}

function isLiteral(expr: Expr): boolean {
  return expr.kind === "IntLit" || expr.kind === "FloatLit" || expr.kind === "StringLit";
}

function literalValue(expr: Expr): string {
  switch (expr.kind) {
    case "IntLit":
      return String(expr.value);
    case "FloatLit":
      return String(expr.value);
    case "StringLit":
      return `'${escapeSqlString(expr.value)}'`;
    default:
      return "NULL";
  }
}

function extractInvariantChecks(inv: Expr, fields: readonly FieldDecl[]): string[] {
  const checks: string[] = [];

  if (inv.kind === "BinaryOp" && inv.op === "and") {
    checks.push(...extractInvariantChecks(inv.left, fields));
    checks.push(...extractInvariantChecks(inv.right, fields));
    return checks;
  }

  if (inv.kind === "BinaryOp") {
    const check = tryMapInvariantComparison(inv, fields);
    if (check) checks.push(check);
  }

  if (inv.kind === "Matches") {
    const fieldName = extractFieldName(inv.expr);
    if (fieldName) {
      checks.push(`${toColumnName(fieldName)} ~ '${escapeSqlString(inv.pattern)}'`);
    }
  }

  if (inv.kind === "BinaryOp" && inv.op === "in" && inv.right.kind === "SetLiteral") {
    const fieldName = extractFieldName(inv.left);
    if (fieldName) {
      const values = inv.right.elements
        .map((e) => {
          if (e.kind === "StringLit") return `'${escapeSqlString(e.value)}'`;
          if (e.kind === "Identifier") return `'${escapeSqlString(e.name)}'`;
          if (e.kind === "EnumAccess") return `'${escapeSqlString(e.member)}'`;
          return null;
        })
        .filter(Boolean);
      if (values.length > 0) {
        checks.push(`${toColumnName(fieldName)} IN (${values.join(", ")})`);
      }
    }
  }

  return checks;
}

function tryMapInvariantComparison(
  expr: Expr & { kind: "BinaryOp" },
  fields: readonly FieldDecl[],
): string | null {
  const sqlOps: Record<string, string> = {
    ">": ">",
    "<": "<",
    ">=": ">=",
    "<=": "<=",
    "=": "=",
    "!=": "!=",
  };

  const sqlOp = sqlOps[expr.op];
  if (!sqlOp) return null;

  if (isLenCall(expr.left) && isLiteral(expr.right)) {
    const arg = (expr.left as { kind: "Call"; args: readonly Expr[] }).args[0];
    const fieldName = extractFieldName(arg);
    if (fieldName && fields.some((f) => f.name === fieldName)) {
      return `length(${toColumnName(fieldName)}) ${sqlOp} ${literalValue(expr.right)}`;
    }
  }

  const leftField = extractFieldName(expr.left);
  if (leftField && isLiteral(expr.right) && fields.some((f) => f.name === leftField)) {
    return `${toColumnName(leftField)} ${sqlOp} ${literalValue(expr.right)}`;
  }

  return null;
}

function extractFieldName(expr: Expr): string | null {
  if (expr.kind === "Identifier") return expr.name;
  if (expr.kind === "FieldAccess" && expr.base.kind === "Identifier") return expr.field;
  return null;
}

File: modules/convention/src/main/scala/specrest/convention/Schema.scala
package specrest.convention
import specrest.ir.Naming
import specrest.ir.generated.SpecRestGenerated.*
import specrest.ir.idx

object Schema:

  // An explicit integer `id` PK is widened to 64-bit. Spec integer semantics are
  // unbounded, so a 32-bit INTEGER overflows on ids >= 2^31 — surfacing as a 500
  // instead of a 4xx not-found — and it would also disagree with synthesized
  // BIGSERIAL PKs and BIGINT FKs. This is the single source of truth: both the
  // migration DDL (deriveTable) and the codegen renderer types (profile.Annotate,
  // which drives Prisma/SQLAlchemy/Go) must apply the *same* rule or the generated
  // client is mistyped against its own migration.

  final private case class EntityRef(tableName: String, idFkSqlType: String)

  final private case class MappedField(
      column: column_spec,
      foreignKey: Option[foreign_key_spec],
      check: Option[String]
  )

  final private case class ClassifierCtx(
      aliasAList: List[(String, type_alias_decl_full)],
      enumAList: List[(String, enum_decl_full)],
      entityNamesList: List[String]
  )

  def deriveSchema(ir: ServiceIRFull): database_schema =
    val ix          = ir.idx
    val entities    = ix.entities
    val entityNames = ix.entityNames
    val cctx = ClassifierCtx(
      aliasAList = ix.aliasAList,
      enumAList = ix.enumAList,
      entityNamesList = ix.entityNamesList
    )
    val entityRefs = buildEntityRefMap(ir, cctx)

    val tables = List.newBuilder[table_spec]
    for entity <- entities do
      tables += deriveTable(entity, ir, entityNames, cctx, entityRefs)

    svcState(ir) match
      case Some(sd) =>
        for sf <- stdFields(sd) do
          stfType(sf) match
            case RelationTypeF(_, _: (MultSome | MultSet), _, _) =>
              deriveJunctionTable(sf, entityNames).foreach(tables += _)
            case _ => ()
      case None => ()

    val builtTables        = tables.result()
    val withPartialIndexes = applyPartialIndexConventions(builtTables, entities, svcConventions(ir))
    val triggers           = detectAggregateTriggers(entities, withPartialIndexes)
    DatabaseSchema(withPartialIndexes, triggers)

  private def buildEntityRefMap(
      ir: ServiceIRFull,
      cctx: ClassifierCtx
  ): Map[String, EntityRef] =
    ir.idx.entities.map { entity =>
      val fields = entFields(entity)
      val tableName = Path
        .getConvention(svcConventions(ir), entName(entity), "db_table")
        .getOrElse(Naming.toTableName(entName(entity)))
      val idField = fields.find(f => fldName(f) == "id")
      val idFkSqlType = idField match
        case None => "BIGINT"
        case Some(f) =>
          val mapped = mapTypeToColumn("id", fldType(f), cctx)
          // A FK referencing the PK must match its widened type.
          columnSqlType(mapped.column) match
            case "BIGSERIAL" => "BIGINT"
            case other       => widenExplicitIdPkSqlType("id", other)
      entName(entity) -> EntityRef(tableName, idFkSqlType)
    }.toMap

  private def deriveTable(
      entity: entity_decl_full,
      ir: ServiceIRFull,
      entityNames: Set[String],
      cctx: ClassifierCtx,
      entityRefs: Map[String, EntityRef]
  ): table_spec =
    val fields = entFields(entity)
    val tableName = Path
      .getConvention(svcConventions(ir), entName(entity), "db_table")
      .getOrElse(Naming.toTableName(entName(entity)))

    val entityFieldNames = fields.map(fldName).toSet
    val columns          = scala.collection.mutable.ListBuffer.empty[column_spec]
    if !entityFieldNames.contains("id") then
      columns += ColumnSpec("id", "BIGSERIAL", false, None)

    val foreignKeys = scala.collection.mutable.ListBuffer.empty[foreign_key_spec]
    val checks      = scala.collection.mutable.ListBuffer.empty[String]
    val indexes     = scala.collection.mutable.ListBuffer.empty[index_spec]

    for field <- fields do
      val colName    = Naming.toColumnName(fldName(field))
      val mapped     = mapFieldToColumn(field, cctx, entityRefs)
      val widenedSql = widenExplicitIdPkSqlType(fldName(field), columnSqlType(mapped.column))
      val column = ColumnSpec(
        columnName(mapped.column),
        widenedSql,
        columnNullable(mapped.column),
        columnDefaultValue(mapped.column)
      )
      columns += column
      mapped.foreignKey.foreach: fk =>
        foreignKeys += fk
        val fkCol = fkColumn(fk)
        indexes += IndexSpec(s"idx_${tableName}_$fkCol", List(fkCol), false, None)
      mapped.check.foreach(checks += _)
      fldDefault(field).foreach: c =>
        checks ++= extractChecks(colName, c)
      for refinement <- aliasRefinements(fldType(field), cctx.aliasAList) do
        checks ++= extractChecks(colName, refinement)

    for inv <- entInvariants(entity) do
      checks ++= extractInvariantChecks(inv, fields)

    svcState(ir) match
      case Some(sd) =>
        for sf <- stdFields(sd) do
          stfType(sf) match
            case RelationTypeF(from, mult, to, _)
                if typeName(to).contains(entName(entity)) && (mult match
                  case _: (MultSome | MultSet) => false; case _ => true
                ) =>
              typeName(from).filter(entityNames.contains) match
                case Some(fromName) =>
                  val fkCol =
                    Naming.toColumnName(
                      fromName.replaceAll("([A-Z])", "_$1").toLowerCase.stripPrefix("_")
                    ) + "_id"
                  val fkRefTable = Naming.toTableName(fromName)
                  val nullable = mult match
                    case _: MultLone => true
                    case _           => false
                  if !columns.exists(c => columnName(c) == fkCol) then
                    columns += ColumnSpec(fkCol, "BIGINT", nullable, None)
                    foreignKeys += ForeignKeySpec(fkCol, fkRefTable, "id", "CASCADE")
                    indexes += IndexSpec(
                      s"idx_${tableName}_$fkCol",
                      List(fkCol),
                      false,
                      None
                    )
                case None => ()
            case _ => ()
      case None => ()

    val tsOverride    = Path.getConvention(svcConventions(ir), entName(entity), "db_timestamps")
    val addTimestamps = !tsOverride.contains("false")
    if addTimestamps && !columns.exists(c => columnName(c) == "created_at") then
      columns += ColumnSpec("created_at", "TIMESTAMPTZ", false, Some("NOW()"))
    if addTimestamps && !columns.exists(c => columnName(c) == "updated_at") then
      columns += ColumnSpec("updated_at", "TIMESTAMPTZ", false, Some("NOW()"))

    TableSpec(
      tableName,
      entName(entity),
      columns.toList,
      "id",
      foreignKeys.toList,
      checks.toList,
      indexes.toList
    )

  private def deriveJunctionTable(
      stateField: state_field_decl_full,
      entityNames: Set[String]
  ): Option[table_spec] =
    stfType(stateField) match
      case RelationTypeF(from, _, to, _) =>
        for
          fromName <- typeName(from) if entityNames.contains(fromName)
          toName   <- typeName(to) if entityNames.contains(toName)
        yield
          val fromTable = Naming.toTableName(fromName)
          val toTable   = Naming.toTableName(toName)
          val tableName = s"${fromTable}_$toTable"
          val fromCol = Naming.toColumnName(
            fromName.replaceAll("([A-Z])", "_$1").toLowerCase.stripPrefix("_")
          ) + "_id"
          val toCol = Naming.toColumnName(
            toName.replaceAll("([A-Z])", "_$1").toLowerCase.stripPrefix("_")
          ) + "_id"
          TableSpec(
            tableName,
            s"${fromName}_$toName",
            List(
              ColumnSpec("id", "BIGSERIAL", false, None),
              ColumnSpec(fromCol, "BIGINT", false, None),
              ColumnSpec(toCol, "BIGINT", false, None),
              ColumnSpec("created_at", "TIMESTAMPTZ", false, Some("NOW()"))
            ),
            "id",
            List(
              ForeignKeySpec(fromCol, fromTable, "id", "CASCADE"),
              ForeignKeySpec(toCol, toTable, "id", "CASCADE")
            ),
            Nil,
            List(
              IndexSpec(
                s"idx_${tableName}_${fromCol}_$toCol",
                List(fromCol, toCol),
                true,
                None
              ),
              IndexSpec(s"idx_${tableName}_$fromCol", List(fromCol), false, None),
              IndexSpec(s"idx_${tableName}_$toCol", List(toCol), false, None)
            )
          )
      case _ => None

  private def mapFieldToColumn(
      field: field_decl_full,
      cctx: ClassifierCtx,
      entityRefs: Map[String, EntityRef]
  ): MappedField =
    val colName = Naming.toColumnName(fldName(field))
    val mapped  = mapTypeToColumn(colName, fldType(field), cctx)
    if mapped.foreignKey.isEmpty && colName.endsWith("_id") then
      val prefix       = colName.dropRight("_id".length)
      val targetEntity = cctx.entityNamesList.find(n => Naming.toSnakeCase(n) == prefix)
      targetEntity.flatMap(entityRefs.get) match
        case Some(ref) =>
          val widened = ColumnSpec(
            columnName(mapped.column),
            ref.idFkSqlType,
            columnNullable(mapped.column),
            columnDefaultValue(mapped.column)
          )
          MappedField(
            column = widened,
            foreignKey = Some(ForeignKeySpec(colName, ref.tableName, "id", "CASCADE")),
            check = mapped.check
          )
        case None => mapped
    else mapped

  private def mapTypeToColumn(
      colName: String,
      typeExpr: type_expr_full,
      cctx: ClassifierCtx
  ): MappedField =
    val classified =
      classifyColumnType(typeExpr, cctx.aliasAList, cctx.enumAList, cctx.entityNamesList)
    val (kind, nullable) = classified match
      case ClassifiedColumn(k, n) => (k, n)
    kind match
      case CkPrim(sqlType) =>
        MappedField(ColumnSpec(colName, sqlType, nullable, None), None, None)
      case CkEnum(vs) =>
        val values = vs.map(v => s"'${escapeSqlString(v)}'").mkString(", ")
        MappedField(
          ColumnSpec(colName, "TEXT", nullable, None),
          None,
          Some(s"$colName IN ($values)")
        )
      case CkEntityRef(name) =>
        val refTable = Naming.toTableName(name)
        val fkCol    = colName + "_id"
        MappedField(
          ColumnSpec(fkCol, "BIGINT", nullable, None),
          Some(ForeignKeySpec(fkCol, refTable, "id", "CASCADE")),
          None
        )
      case _: CkJsonArray =>
        MappedField(ColumnSpec(colName, "JSONB", nullable, Some("'[]'::jsonb")), None, None)
      case _: CkJsonObject =>
        MappedField(ColumnSpec(colName, "JSONB", nullable, Some("'{}'::jsonb")), None, None)
      case _: CkRelation =>
        MappedField(ColumnSpec(colName, "BIGINT", nullable, None), None, None)
      case _: CkUnknown =>
        MappedField(ColumnSpec(colName, "TEXT", nullable, None), None, None)

  private def escapeSqlString(s: String): String = s.replace("'", "''")

  private def extractChecks(colName: String, constraint: expr_full): List[String] =
    flattenAnd(constraint).flatMap: atom =>
      classifyColumnCheckAtom(atom) match
        case _: CcSkip => Nil
        case CcRegexMatch(pat) =>
          List(s"$colName ~ '${escapeSqlString(pat)}'")
        case CcLenCompare(op, n) =>
          sqlOp(op).toList.map(o => s"length($colName) $o $n")
        case CcValueCompare(op, n) =>
          sqlOp(op).toList.map(o => s"$colName $o $n")
        case CcLenLitCompare(op, rhs) =>
          sqlOp(op).toList.map(o => s"length($colName) $o ${literalValue(rhs)}")
        case CcValueLitCompare(op, rhs) =>
          sqlOp(op).toList.map(o => s"$colName $o ${literalValue(rhs)}")

  private def literalValue(e: expr_full): String = e match
    case IntLitF(v, _)    => v.toString
    case FloatLitF(v, _)  => v
    case StringLitF(v, _) => s"'${escapeSqlString(v)}'"
    case _                => "NULL"

  private def extractInvariantChecks(
      inv: expr_full,
      @annotation.unused fields: List[field_decl_full]
  ): List[String] =
    flattenAnd(inv).flatMap: atom =>
      classifyInvariantAtom(atom) match
        case _: IcSkip => Nil
        case IcInClause(fieldName, elements) =>
          val colName = Naming.toColumnName(fieldName)
          val values = elements.collect {
            case StringLitF(v, _) => s"'${escapeSqlString(v)}'"
            case IntLitF(v, _)    => v.toString
          }
          // Lifted predicate already required all elements to be literals.
          // The Scala collect narrows further to String/Int (Float / Bool /
          // None aren't supported as IN-clause values), so an atom with a
          // non-(String|Int) literal element is dropped here.
          if values.length == elements.length && values.nonEmpty then
            List(s"$colName IN (${values.mkString(", ")})")
          else Nil
        case IcCompare(fieldName, op, rhs) =>
          val colName = Naming.toColumnName(fieldName)
          // sqlOp is total on the operators the lifted classifier admits
          // (Gt/Lt/Ge/Le/Eq/Neq) — see SchemaDerive.sqlOp_def.
          sqlOp(op).toList.map(o => s"$colName $o ${literalValue(rhs)}")

  private def applyPartialIndexConventions(
      tables: List[table_spec],
      entities: List[entity_decl_full],
      conv: Option[conventions_decl_full]
  ): List[table_spec] =
    val rules = extractPartialIndexRules(conv)
    if rules.isEmpty then tables
    else
      val entityByName = entities.map(e => entName(e) -> e).toMap
      // Resolve target entity name → table name (regex + convention lookup, stays in Scala).
      val rulesByTable: Map[String, List[(String, String)]] = rules
        .flatMap { case (target, (col, filt)) =>
          entityByName.get(target).map: e =>
            val tableNm =
              Path.getConvention(conv, entName(e), "db_table").getOrElse(
                Naming.toTableName(entName(e))
              )
            (tableNm, Naming.toColumnName(col), filt)
        }
        .groupBy(_._1)
        .view
        .mapValues(_.map((_, c, f) => (c, f)))
        .toMap
      tables.map: t =>
        rulesByTable.get(tableName(t)) match
          case None             => t
          case Some(colFilters) => appendPartialIndexes(t, colFilters)

  private def detectAggregateTriggers(
      entities: List[entity_decl_full],
      tables: List[table_spec]
  ): List[trigger_spec] =
    val tablesByEntity = tables.map(t => tableEntityName(t) -> t).toMap
    val entitiesByName = entities.map(e => entName(e) -> e).toMap
    val out            = List.newBuilder[trigger_spec]
    for
      parent    <- entities
      parentTbl <- tablesByEntity.get(entName(parent))
      inv       <- entInvariants(parent)
    do
      detectAggregateInvariant(inv) match
        case Some(DetectedAggregate(targetField, collFieldName, agg, sourceField)) =>
          val parentFields = entFields(parent)
          val triggerOpt: Option[trigger_candidate] =
            for
              collField   <- parentFields.find(f => fldName(f) == collFieldName)
              childName   <- collectionElementEntityName(fldType(collField))
              childEntity <- entitiesByName.get(childName)
              childTable  <- tablesByEntity.get(childName)
              validated <- validateTrigger(
                             parentTbl,
                             parentFields,
                             childTable,
                             childEntity,
                             targetField,
                             agg,
                             sourceField
                           )
            yield validated
          triggerOpt match
            case Some(TriggerCandidate(parentTable, _, childTable, fkCol, _, srcField)) =>
              val parentSnake = Naming.toSnakeCase(entName(parent))
              val funcName    = s"recalc_${parentSnake}_$targetField"
              out += TriggerSpec(
                s"trg_$funcName",
                funcName,
                parentTable,
                Naming.toColumnName(targetField),
                childTable,
                fkCol,
                agg,
                srcField.map(Naming.toColumnName)
              )
            case None => ()
        case None => ()
    out.result()
